pkg/syncer/synctarget/shard_manager.go

```diff
@@ -0,0 +1,159 @@
+// cluster-aware informer factories for both the Syncer and Upsyncer virtual workspaces
+// associated to a Shard.
+type ShardAccess struct {
+	SyncerClient   kcpdynamic.ClusterInterface
+	SyncerDDSIF    *informer.DiscoveringDynamicSharedInformerFactory
+	UpsyncerClient kcpdynamic.ClusterInterface
+	UpsyncerDDSIF  *informer.DiscoveringDynamicSharedInformerFactory
+}
+
+// GetShardAccessFunc is the type of a function that provide a [ShardAccess] from a
+// logical cluster name.
+type GetShardAccessFunc func(clusterName logicalcluster.Name) (ShardAccess, error)
+
+// NewShardManager returns a [ShardManager] that can manage the addition or removal of shard-specific
+// upstream virtual workspace URLs, based on a SyncTarget resource passed to the updateShards() method.
+//
+// When a shard is found (identified by the couple of virtual workspace URLs - for both syncer and upsyncer),
+// then the startShardControllers() method is called, and the resulting [ShardAccess] is stored.
+//
+// When a shard is removed, the cleanupShard() method is called, and the context initially passed to the
+// startShardControllers() method is cancelled.
+//
+// The ShardAccessForCluster() method will be used by some downstream controllers in order to
+// be able to get / list upstream resources in the right shard.
+func NewShardManager(
+	gvrSource informer.GVRSource,
+	startShardControllers func(ctx context.Context, gvrSource informer.GVRSource, shardURLs workloadv1alpha1.VirtualWorkspace) (*ShardAccess, error),
+	cleanupShard func(shardURLs workloadv1alpha1.VirtualWorkspace)) *shardManager {
+	return &shardManager{
+		shardsWithStartedControllers: map[workloadv1alpha1.VirtualWorkspace]shardWithStartedControllers{},
+		startShardControllers:        startShardControllers,
+		cleanupShard:                 cleanupShard,
+		gvrSource:                    gvrSource,
+	}
+}
+
+type shardManager struct {
+	shardsWithStartedControllersLock sync.RWMutex
+	shardsWithStartedControllers     map[workloadv1alpha1.VirtualWorkspace]shardWithStartedControllers
+	startShardControllers            func(ctx context.Context, gvrSource informer.GVRSource, shardURLs workloadv1alpha1.VirtualWorkspace) (*ShardAccess, error)
+	cleanupShard                     func(shardURLs workloadv1alpha1.VirtualWorkspace)
+	gvrSource                        informer.GVRSource
+}
+
+func (c *shardManager) ShardAccessForCluster(clusterName logicalcluster.Name) (ShardAccess, error) {
+	c.shardsWithStartedControllersLock.RLock()
+	defer c.shardsWithStartedControllersLock.RUnlock()
+
+	for _, shardAccess := range c.shardsWithStartedControllers {
+		if _, err := shardAccess.logicalClusterLister.ByCluster(clusterName).Get(kcpcorev1alpha1.LogicalClusterName); err == nil {
+			return shardAccess.ShardAccess, nil
+		} else if !apierrors.IsNotFound(err) {
+			return ShardAccess{}, err
+		}
+	}
+	return ShardAccess{}, apierrors.NewNotFound(logicalClusterGVR.GroupResource(), kcpcorev1alpha1.LogicalClusterName)
```

Comment:
don't use api errors for non-API code. `ShardAccess, bool, error` would be canonical.

Reply:
Fixed